File: algorithms/12_php/sorting/CountingSort.php
<?php
declare(strict_types=1);

namespace Algorithms\Sorting;

function countingSort(array $arr): array {
    if (count($arr) === 0) return [];
    $maxVal = max($arr);
    $count = array_fill(0, $maxVal + 1, 0);
    foreach ($arr as $x) $count[$x]++;
    for ($i = 1; $i <= $maxVal; $i++) $count[$i] += $count[$i - 1];
    $out = array_fill(0, count($arr), 0);
    foreach (array_reverse($arr) as $x) {
        $count[$x]--;
        $out[$count[$x]] = $x;
    }
    return $out;
}

if (php_sapi_name() === 'cli' && basename($_SERVER['SCRIPT_FILENAME'] ?? '') === basename(__FILE__)) {
    $data = [4, 2, 2, 8, 3, 3, 1];
    $sorted = countingSort($data);
    $expected = $data;
    sort($expected);
    if ($sorted !== $expected) {
        throw new \RuntimeException("CountingSort verification failed");
    }
    echo "[PHP CountingSort] Counting sort verified: " . json_encode($sorted) . "\n";
}

File: algorithms/05_csharp/sorting/counting_sort.cs
using System;

namespace Algorithms.Sorting
{
    public static class CountingSort
    {
        public static int[] Sort(int[] array)
        {
            if (array.Length == 0) return array;

            int min = array[0], max = array[0];
            foreach (int x in array)
            {
                if (x < min) min = x;
                if (x > max) max = x;
            }

            int range = max - min + 1;
            var count = new int[range];
            foreach (int x in array) count[x - min]++;

            var result = new int[array.Length];
            int idx = 0;
            for (int i = 0; i < range; i++)
            {
                for (int j = 0; j < count[i]; j++)
                    result[idx++] = i + min;
            }
            return result;
        }

        public static void Main()
        {
            var sorted = Sort(new[] { 4, 2, 2, 8, 3, 3, 1 });
            bool ok = sorted[0] == 1 && sorted[1] == 2 && sorted[2] == 2 && sorted[6] == 8;
            Console.WriteLine(ok ? "[C# CountingSort] ... verified" : "[C# CountingSort] ... FAILED");
        }
    }
}
